core/solver/gmres.cpp

```diff
@@ -240,76 +281,89 @@ void Gmres<ValueType>::apply_dense_impl(const matrix::Dense<ValueType>* dense_b,
         // next_krylov = A * preconditioned_vector
         this->get_system_matrix()->apply(preconditioned_vector,
                                          next_krylov.get());
 
-        // final_iter_nums += 1 (unconverged)
-        // next_krylov_basis is alias for (restart_iter + 1)-th krylov_bases
-        // for i in 0:restart_iter(include)
-        //     hessenberg(restart_iter, i) = next_krylov_basis' *
-        //         krylov_bases(:, i)
-        //     next_krylov_basis  -= hessenberg(restart_iter, i) *
-        //         krylov_bases(:, i)
-        // end
-        // hessenberg(restart_iter+1, restart_iter) = norm(next_krylov_basis)
-        // next_krylov_basis /= hessenberg(restart_iter + 1, restart_iter)
-        // End of arnoldi
-        // Start apply givens rotation
+        for (size_type i = 0; i <= restart_iter; i++) {
+            // orthogonalize against krylov_bases(:, i):
+            // hessenberg(i, restart_iter) = next_krylov' * krylov_bases(:, i)
+            // next_krylov -= hessenberg(i, restart_iter) * krylov_bases(:, i)
+            auto hessenberg_entry = hessenberg_iter->create_submatrix(
+                span{i, i + 1}, span{0, num_rhs});
+            auto krylov_basis = krylov_bases->create_submatrix(
+                span{num_rows * i, num_rows * (i + 1)}, span{0, num_rhs});
+            next_krylov->compute_conj_dot(
+                krylov_basis.get(), hessenberg_entry.get(), reduction_tmp);
+            next_krylov->sub_scaled(hessenberg_entry.get(), krylov_basis.get());
+        }
+        // normalize next_krylov:
+        // hessenberg(restart_iter+1, restart_iter) = norm(next_krylov)
+        // next_krylov /= hessenberg(restart_iter+1, restart_iter)
+        auto hessenberg_norm_entry = hessenberg_iter->create_submatrix(
+            span{restart_iter + 1, restart_iter + 2}, span{0, num_rhs});
+        help_compute_norm<ValueType>::compute_next_krylov_norm_into_hessenberg(
+            next_krylov.get(), hessenberg_norm_entry.get(),
+            next_krylov_norm_tmp, reduction_tmp);
+        next_krylov->inv_scale(hessenberg_norm_entry.get());
+        // End of Arnoldi
+
+        // update QR factorization and Krylov RHS for last column:
+        // apply givens rotation
         // for j in 0:restart_iter(exclude)
         //     temp             =  cos(j)*hessenberg(j) +
         //                         sin(j)*hessenberg(j+1)
         //     hessenberg(j+1)  = -conj(sin(j))*hessenberg(j) +
         //                         conj(cos(j))*hessenberg(j+1)
         //     hessenberg(j)    =  temp;
         // end
-        // Calculate sin and cos
+        // calculate next Givens parameters
         // this_hess = hessenberg(restart_iter)
         // next_hess = hessenberg(restart_iter+1)
-        // hypotenuse = sqrt(this_hess * this_hess + next_hess * next_hess);
-        // cos(restart_iter) = conj(this_hess) / hypotenuse;
+        // hypotenuse = norm2([this_hess next_hess])
```

Comment:
```suggestion
        // hypotenuse = ||(this_hess, next_hess)||
```
using (,) as point position (x, y).